File: SOURCE_NATIVE/EW4_iOS_NativePort/NativeCore/Sources/EW4NativeCore/CombatCore.swift
import Foundation

public struct CombatUnitSnapshot: Sendable {
    public var grade: Int
    public var hp: Int
    public var maxHP: Int
    public var forceFullFormation: Bool
    public init(grade:Int,hp:Int,maxHP:Int,forceFullFormation:Bool=false){self.grade=grade;self.hp=hp;self.maxHP=maxHP;self.forceFullFormation=forceFullFormation}
}

public struct CombatStat: Sendable {
    public var type: String
    public var minAttack: Int
    public var maxAttack: Int
    public init(type:String,minAttack:Int=1,maxAttack:Int=1){self.type=type;self.minAttack=minAttack;self.maxAttack=maxAttack}
}

public struct CombatCommander: Sendable {
    public var infantry:Int=0,cavalry:Int=0,artillery:Int=0,warship:Int=0,fort:Int=0
    public var skills:Set<Int>=[]
    public init(infantry:Int=0,cavalry:Int=0,artillery:Int=0,warship:Int=0,fort:Int=0,skills:Set<Int>=[]){self.infantry=infantry;self.cavalry=cavalry;self.artillery=artillery;self.warship=warship;self.fort=fort;self.skills=skills}
}

public struct AttackInterval: Equatable, Sendable { public let min:Int; public let max:Int }
public struct SkillIntervalBonus: Equatable, Sendable { public let lower:Int; public let upper:Int }
public struct DamageBounds: Sendable {
    public let min,max,attackMin,attackMax,mastery,coefficient,masteryCoefficient,fixedBonus,skillLower,skillUpper:Int
}
public struct DamageResult: Sendable {
    public let bounds:DamageBounds
    public let value:Int
    public let attackTactic:Bool
    public let defenseTactic:Bool
    public let totalReduction:Int
    public let notes:[String]
}

public struct DamageOptions: Sendable {
    public var attacker:CombatUnitSnapshot
    public var attackerStat:CombatStat
    public var attackerCommander:CombatCommander?
    public var defenderStat:CombatStat
    public var defenderCommander:CombatCommander?
    public var isCounter=false
    public var isPlayerAttacker=false
    public var attackerIsFort=false
    public var defenderIsFort=false
    public var fixedBonus=0,flatBonus=0,lowerBonus=0,upperBonus=0
    public var attackerMorale=0,defenderMorale=0,defenderTrainingDefense=0
    public var attackerEmbarked=false
    public var terrainReduction=0,installationReduction=0,buildingReduction=0,countryReduction=0
    public var defenderHasConstruction=false
    public init(attacker:CombatUnitSnapshot,attackerStat:CombatStat,attackerCommander:CombatCommander?=nil,defenderStat:CombatStat=CombatStat(type:"infantry"),defenderCommander:CombatCommander?=nil){self.attacker=attacker;self.attackerStat=attackerStat;self.attackerCommander=attackerCommander;self.defenderStat=defenderStat;self.defenderCommander=defenderCommander}
}

public enum CombatCore {
    public static let playerAttackIntervalBonus=4
    public static let attackTacticChance=0.10
    public static let defenseTacticChance=0.10

    private static func clamp(_ v:Int,_ a:Int,_ b:Int)->Int{max(a,min(b,v))}
    public static func hasSkill(_ commander:CombatCommander?,_ id:Int)->Bool{commander?.skills.contains(id) ?? false}

    public static func skillIntervalBonus(_ commander:CombatCommander?, statType:String)->SkillIntervalBonus{
        var lower=0,upper=0
        switch statType {
        case "infantry": if hasSkill(commander,13){lower += 1}; if hasSkill(commander,12){upper += 1}
        case "cavalry": if hasSkill(commander,11){lower += 1}; if hasSkill(commander,10){upper += 1}
        case "artillery": if hasSkill(commander,7){lower += 1}; if hasSkill(commander,8){upper += 1}
        default: break
        }
        return .init(lower:lower,upper:upper)
    }

    public static func effectiveAttackInterval(min baseMin:Int,max baseMax:Int,isPlayer:Bool=false,flatBonus:Int=0,lowerBonus:Int=0,upperBonus:Int=0)->AttackInterval{
        let player=isPlayer ? playerAttackIntervalBonus : 0
        let lo=baseMin+flatBonus+lowerBonus+player
        let rawHi=baseMax+flatBonus+upperBonus+player
        return .init(min:lo,max:max(lo,rawHi))
    }

    public static func masteryBonus(_ commander:CombatCommander?,statType:String,isFort:Bool=false)->Int{
        guard let c=commander else{return 0};if isFort{return c.fort}
        switch statType{case"infantry":return c.infantry;case"cavalry":return c.cavalry;case"artillery":return c.artillery;case"warship":return c.warship;default:return 0}
    }
    public static func fullFormationCoefficient(_ unit:CombatUnitSnapshot)->Int{clamp(5+clamp(unit.grade,0,2),5,7)}
    public static func formationCoefficient(_ unit:CombatUnitSnapshot,_ commander:CombatCommander?,statType:String)->Int{
        let full=fullFormationCoefficient(unit)
        if unit.forceFullFormation || (statType=="infantry" && hasSkill(commander,15)){return full}
        let ratio=Double(max(0,unit.hp))/Double(max(1,unit.maxHP))
        if full>=7 && ratio>=0.80{return 7};if full>=6 && ratio>=0.65{return 6};if ratio>=0.50{return 5};if ratio>=0.30{return 4};if ratio>=0.15{return 3};if ratio>=0.05{return 2};return 1
    }
    public static func masteryCoefficient(_ unit:CombatUnitSnapshot,_ commander:CombatCommander?,statType:String)->Int{
        if unit.forceFullFormation || (statType=="infantry" && hasSkill(commander,15)){return 5}
        let ratio=Double(max(0,unit.hp))/Double(max(1,unit.maxHP));if ratio>=0.50{return 5};if ratio>=0.30{return 4};if ratio>=0.15{return 3};if ratio>=0.05{return 2};return 1
    }

    public static func damageBounds(unit:CombatUnitSnapshot,stat:CombatStat,commander:CombatCommander?=nil,isPlayer:Bool=false,isFort:Bool=false,fixedBonus:Int=0,flatBonus:Int=0,lowerBonus:Int=0,upperBonus:Int=0)->DamageBounds{
        let skill=skillIntervalBonus(commander,statType:stat.type)
        let interval=effectiveAttackInterval(min:stat.minAttack,max:stat.maxAttack,isPlayer:isPlayer,flatBonus:flatBonus,lowerBonus:lowerBonus+skill.lower,upperBonus:upperBonus+skill.upper)
        let mastery=masteryBonus(commander,statType:stat.type,isFort:isFort), coeff=formationCoefficient(unit,commander,statType:stat.type), masteryCoeff=masteryCoefficient(unit,commander,statType:stat.type)
        let fixed=mastery*masteryCoeff+fixedBonus
        let minDamage=max(1,fixed+coeff*interval.min),maxDamage=max(minDamage,fixed+coeff*interval.max)
        return .init(min:minDamage,max:maxDamage,attackMin:interval.min,attackMax:interval.max,mastery:mastery,coefficient:coeff,masteryCoefficient:masteryCoeff,fixedBonus:fixedBonus,skillLower:skill.lower,skillUpper:skill.upper)
    }

    private static func rollInclusive(_ minValue:Int,_ maxValue:Int,_ rng:()->Double)->Int{
        let lo=minValue,hi=max(lo,maxValue),r=min(max(rng(),0),0.999999999999);return lo+Int(floor(r*Double(hi-lo+1)))
    }
    public static func rollDamage(unit:CombatUnitSnapshot,stat:CombatStat,commander:CombatCommander?=nil,isPlayer:Bool=false,isFort:Bool=false,fixedBonus:Int=0,flatBonus:Int=0,lowerBonus:Int=0,upperBonus:Int=0,rng:()->Double={Double.random(in:0..<1)},forceMax:Bool=false)->(bounds:DamageBounds,value:Int){
        let b=damageBounds(unit:unit,stat:stat,commander:commander,isPlayer:isPlayer,isFort:isFort,fixedBonus:fixedBonus,flatBonus:flatBonus,lowerBonus:lowerBonus,upperBonus:upperBonus)
        var dice=0;if forceMax{dice=b.coefficient*b.attackMax}else{for _ in 0..<b.coefficient{dice += rollInclusive(b.attackMin,b.attackMax,rng)}}
        let fixed=b.mastery*b.masteryCoefficient+b.fixedBonus;return(b,max(1,fixed+dice))
    }
    public static func ignoresAvoidance(_ commander:CombatCommander?,statType:String)->Bool{statType=="infantry" ? hasSkill(commander,14) : statType=="cavalry" ? hasSkill(commander,9) : statType=="artillery" ? hasSkill(commander,5) : false}

    public static func resolveDamage(_ options:DamageOptions,rng:()->Double={Double.random(in:0..<1)})->DamageResult{
        let aCmd=options.attackerCommander,dCmd=options.defenderCommander,aStat=options.attackerStat,dStat=options.defenderStat
        let attackTactic=hasSkill(aCmd,31) && !options.isCounter && rng()<attackTacticChance
        let rolled=rollDamage(unit:options.attacker,stat:aStat,commander:aCmd,isPlayer:options.isPlayerAttacker,isFort:options.attackerIsFort,fixedBonus:options.fixedBonus,flatBonus:options.flatBonus,lowerBonus:options.lowerBonus,upperBonus:options.upperBonus,rng:rng,forceMax:attackTactic)
        var value=rolled.value,notes:[String]=[];if attackTactic{notes.append("attack-tactics")}
        value += options.attackerMorale*rolled.bounds.coefficient;if options.defenderMorale<0{value += (-options.defenderMorale)*rolled.bounds.coefficient}
        if options.defenderTrainingDefense>0{value=max(1,value-options.defenderTrainingDefense);notes.append("training-defense-\(options.defenderTrainingDefense)")}
        if options.attackerEmbarked && aStat.type != "warship" && !hasSkill(aCmd,17){value=Int(floor(Double(value)*0.8));notes.append("embarked-penalty")}
        if options.defenderIsFort && hasSkill(aCmd,4){value=Int(floor(Double(value)*1.5));notes.append("spy+50%")}
        var terrain=max(0,options.terrainReduction),installation=max(0,options.installationReduction),building=max(0,options.buildingReduction),country=max(0,options.countryReduction)
        if ignoresAvoidance(aCmd,statType:aStat.type){terrain=0;installation=0;building=0;country=0;notes.append("ignore-avoidance")}else if hasSkill(aCmd,29) && building>0{building=0;notes.append("siegecraft")}
        let mapReduction=options.defenderHasConstruction ? building : max(terrain,installation),totalReduction=clamp(mapReduction+country,0,95)
        if totalReduction>0{value=Int(floor(Double(value)*Double(100-totalReduction)/100.0))}
        if dStat.type=="warship" && hasSkill(dCmd,16){value=Int(floor(Double(value)*0.9));notes.append("helmsman-10%")}
        value=max(1,value)
        let defenseTactic=hasSkill(dCmd,32) && !options.isCounter && rng()<defenseTacticChance
        if defenseTactic{value=1;notes.append("defense-tactics")}
        return .init(bounds:rolled.bounds,value:value,attackTactic:attackTactic,defenseTactic:defenseTactic,totalReduction:totalReduction,notes:notes)
    }
}

File: SOURCE_LEAN/EW4_Web_Port_v0.61/combat_core.js
'use strict';
/* EW4 Web Port 0.61 combat core.
   Goals:
   1) preserve EW4's dice/formation damage structure;
   2) FIX the original lower-bound regression: lower attack is a real die floor;
   3) support native skill semantics used by player-enhanced generals;
   4) keep player-only +4/+4 isolated from AI units. */
(function(root,factory){
  const api=factory();
  if(typeof module==='object'&&module.exports)module.exports=api;
  else root.EW4Combat=api;
})(typeof globalThis!=='undefined'?globalThis:this,function(){
  const PLAYER_ATTACK_INTERVAL_BONUS=4;
  const ATTACK_TACTIC_CHANCE=0.10;
  const DEFENSE_TACTIC_CHANCE=0.10;
  const clamp=(v,a,b)=>Math.max(a,Math.min(b,v));
  const num=(v,d=0)=>Number.isFinite(+v)?+v:d;

  function normalizeInterval(min,max){
    let lo=num(min,0),hi=num(max,lo);
    if(hi<lo)hi=lo; // a valid lower-bound increase is NEVER deleted
    return {min:lo,max:hi};
  }

  function hasSkill(commander,id){
    if(!commander)return false;
    id=+id;
    const legacy=[commander.skill1,commander.skill2,commander.skill3,commander.skill4];
    const extended=Array.isArray(commander.skills)?commander.skills:[];
    return [...legacy,...extended].some(x=>+x===id);
  }

  function skillIntervalBonus(commander,statType){
    let lower=0,upper=0;
    // Native EW4 intent. The original game failed to use these lower floors correctly;
    // 0.61 intentionally repairs that regression.
    if(statType==='infantry'){
      if(hasSkill(commander,13))lower+=1; // Infantry Tactics
      if(hasSkill(commander,12))upper+=1; // Formation
    }else if(statType==='cavalry'){
      if(hasSkill(commander,11))lower+=1; // Maneuver
      if(hasSkill(commander,10))upper+=1; // Surprise
    }else if(statType==='artillery'){
      if(hasSkill(commander,7))lower+=1;  // Ballistics
      if(hasSkill(commander,8))upper+=1;  // Explosives
    }
    return {lower,upper};
  }

  function effectiveAttackInterval(opts={}){
    const isPlayer=!!opts.isPlayer;
    const baseMin=num(opts.min,1),baseMax=num(opts.max,baseMin);
    const flat=num(opts.flatBonus,0);
    const lower=num(opts.lowerBonus,0);
    const upper=num(opts.upperBonus,0);
    const player=isPlayer?PLAYER_ATTACK_INTERVAL_BONUS:0;
    return normalizeInterval(baseMin+flat+lower+player,baseMax+flat+upper+player);
  }

  function masteryBonus(commander,statType,isFort=false){
    if(!commander)return 0;
    if(isFort)return num(commander.fort,0);
    if(statType==='infantry')return num(commander.infantry,0);
    if(statType==='cavalry')return num(commander.cavalry,0);
    if(statType==='artillery')return num(commander.artillery,0);
    if(statType==='warship')return num(commander.warship,0);
    return 0;
  }

  function fullFormationCoefficient(unit){
    // grade 0/1/2 is EW4 single/double/triple formation.
    return clamp(5+Math.max(0,Math.min(2,Math.trunc(num(unit?.grade,0)))),5,7);
  }

  function formationCoefficient(unit,commander,statType){
    const full=fullFormationCoefficient(unit);
    if(unit?.forceFullFormation||(statType==='infantry'&&hasSkill(commander,15)))return full; // Snare Drum / Mass Fire
    const hp=Math.max(0,num(unit?.hp,0)),max=Math.max(1,num(unit?.max_hp,1));
    const r=hp/max;
    // Recovered EW4 HP breakpoints:
    // single: >=50/30/15/5% => 5/4/3/2, else 1
    // double adds coeff 6 above 65%; triple adds 7 above 80% and 6 above 65%.
    if(full>=7&&r>=0.80)return 7;
    if(full>=6&&r>=0.65)return 6;
    if(r>=0.50)return 5;
    if(r>=0.30)return 4;
    if(r>=0.15)return 3;
    if(r>=0.05)return 2;
    return 1;
  }

  function masteryCoefficient(unit,commander,statType){
    // EW4 commander-star damage decays with the same 50/30/15/5% breakpoints as a
    // single formation. Dense Attack (infantry) and Snare Drum preserve full output.
    if(unit?.forceFullFormation||(statType==='infantry'&&hasSkill(commander,15)))return 5;
    const hp=Math.max(0,num(unit?.hp,0)),max=Math.max(1,num(unit?.max_hp,1)),r=hp/max;
    if(r>=0.50)return 5;if(r>=0.30)return 4;if(r>=0.15)return 3;if(r>=0.05)return 2;return 1;
  }

  function rollInclusive(min,max,rng=Math.random){
    const lo=Math.ceil(num(min,0)),hi=Math.max(lo,Math.floor(num(max,lo)));
    const r=clamp(num(rng(),0),0,0.999999999999);
    return lo+Math.floor(r*(hi-lo+1));
  }

  function damageBounds({unit,stat,commander,isPlayer=false,isFort=false,fixedBonus=0,flatBonus=0,lowerBonus=0,upperBonus=0}={}){
    stat=stat||{};unit=unit||{};
    const s=skillIntervalBonus(commander,stat.type);
    const interval=effectiveAttackInterval({
      min:stat.minatk,max:stat.maxatk,isPlayer,
      flatBonus:num(flatBonus,0),
      lowerBonus:num(lowerBonus,0)+s.lower,
      upperBonus:num(upperBonus,0)+s.upper
    });
    const mastery=masteryBonus(commander,stat.type,isFort);
    const coeff=formationCoefficient(unit,commander,stat.type),masteryCoeff=masteryCoefficient(unit,commander,stat.type);
    // Commander-star contribution also loses tiers as HP falls. Equipment attack remains fixed.
    const fixed=Math.round(mastery*masteryCoeff+num(fixedBonus,0));
    const minDamage=Math.max(1,fixed+coeff*interval.min);
    const maxDamage=Math.max(minDamage,fixed+coeff*interval.max);
    return {min:minDamage,max:maxDamage,attackMin:interval.min,attackMax:interval.max,mastery,coefficient:coeff,masteryCoefficient:masteryCoeff,fixedBonus:num(fixedBonus,0),skillLower:s.lower,skillUpper:s.upper};
  }

  function rollDamage(args={},rng=Math.random,forceMax=false){
    const b=damageBounds(args);
    let dice=0;
    if(forceMax)dice=b.coefficient*b.attackMax;
    else for(let i=0;i<b.coefficient;i++)dice+=rollInclusive(b.attackMin,b.attackMax,rng);
    const fixed=b.mastery*b.masteryCoefficient+b.fixedBonus;
    const value=Math.max(1,Math.round(fixed+dice));
    return {...b,value};
  }

  function ignoresAvoidance(commander,statType){
    return statType==='infantry'?hasSkill(commander,14):
           statType==='cavalry'?hasSkill(commander,9):
           statType==='artillery'?hasSkill(commander,5):false;
  }

  function resolveDamage(opts={},rng=Math.random){
    const aStat=opts.attackerStat||opts.stat||{};
    const dStat=opts.defenderStat||{};
    const aCmd=opts.attackerCommander||opts.commander||null;
    const dCmd=opts.defenderCommander||null;
    const isCounter=!!opts.isCounter;

    // Attack Tactics only applies to the active attack, never to an off-turn counterattack.
    const attackTactic=hasSkill(aCmd,31)&&!isCounter&&rng()<ATTACK_TACTIC_CHANCE;
    const rolled=rollDamage({
      unit:opts.attacker||opts.unit,stat:aStat,commander:aCmd,
      isPlayer:!!opts.isPlayerAttacker,isFort:!!opts.attackerIsFort,
      fixedBonus:num(opts.fixedBonus,0),flatBonus:num(opts.flatBonus,0),
      lowerBonus:num(opts.lowerBonus,0),upperBonus:num(opts.upperBonus,0)
    },rng,attackTactic);

    let value=rolled.value;
    const notes=[];
    if(attackTactic)notes.push('attack-tactics');

    // Morale is a fixed modifier based on the current formation coefficient.
    // +1 morale => +k damage; -1/-2 => -k/-2k. Low defender morale increases incoming by k/2k.
    const aMorale=Math.trunc(num(opts.attackerMorale,0));
    const dMorale=Math.trunc(num(opts.defenderMorale,0));
    value+=aMorale*rolled.coefficient;
    if(dMorale<0)value+=(-dMorale)*rolled.coefficient;

    // Native unit training level contributes a flat defender-side subtraction before
    // subsequent multiplicative combat modifiers. Level 0..5 => 0/2/4/6/8/10.
    const trainingDefense=Math.max(0,Math.trunc(num(opts.defenderTrainingDefense,0)));
    if(trainingDefense>0){value=Math.max(1,value-trainingDefense);notes.push(`training-defense-${trainingDefense}`);}

    // Sailor cancels the 20% attack penalty for land units fighting while embarked.
    if(opts.attackerEmbarked&&aStat.type!=='warship'&&!hasSkill(aCmd,17)){
      value=Math.floor(value*0.8);notes.push('embarked-penalty');
    }

    // Spy: +50% against fortress-family units.
    if(opts.defenderIsFort&&hasSkill(aCmd,4)){
      value=Math.floor(value*1.5);notes.push('spy+50%');
    }

    // Terrain / field-work / building avoidance. Infantry/Cavalry/Artillery ignore skills
    // wipe the known avoidance layers; Siegecraft ignores building avoidance only.
    let terrain=Math.max(0,num(opts.terrainReduction,0));
    let installation=Math.max(0,num(opts.installationReduction,0));
    let building=Math.max(0,num(opts.buildingReduction,0));
    let country=Math.max(0,num(opts.countryReduction,0));
    if(ignoresAvoidance(aCmd,aStat.type)){
      terrain=installation=building=country=0;notes.push('ignore-avoidance');
    }else if(hasSkill(aCmd,29)&&building>0){
      building=0;notes.push('siegecraft');
    }
    // Native defense selector: a construction owns the cell's map-avoidance path.
    // Only a cell without a construction falls back to max(terrain, field installation).
    // This matters even for farmland (construction present, avoid=0): terrain is not added back in.
    const hasConstruction=!!opts.defenderHasConstruction;
    const mapReduction=hasConstruction?building:Math.max(terrain,installation);
    const totalReduction=clamp(mapReduction+country,0,95);
    if(totalReduction>0)value=Math.floor(value*(100-totalReduction)/100);

    // Helmsman: naval unit suffers 10% less damage.
    if(dStat.type==='warship'&&hasSkill(dCmd,16)){
      value=Math.floor(value*0.9);notes.push('helmsman-10%');
    }

    value=Math.max(1,Math.floor(value));

    // Defense Tactics only applies while defending against an active enemy attack.
    // In EW4 it does not trigger when the unit is taking a counterattack on its own turn.
    const defenseTactic=hasSkill(dCmd,32)&&!isCounter&&rng()<DEFENSE_TACTIC_CHANCE;
    if(defenseTactic){value=1;notes.push('defense-tactics');}

    return {...rolled,value,attackTactic,defenseTactic,totalReduction,notes};
  }

  return Object.freeze({
    PLAYER_ATTACK_INTERVAL_BONUS,ATTACK_TACTIC_CHANCE,DEFENSE_TACTIC_CHANCE,
    effectiveAttackInterval,skillIntervalBonus,damageBounds,rollDamage,resolveDamage,
    hasSkill,masteryBonus,masteryCoefficient,formationCoefficient,fullFormationCoefficient,ignoresAvoidance
  });
});
